### PR TITLE
Run the ProcessExit events in the threadpool and enforce a 3 seconds limit. Fixes #10087

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

File: mono/metadata/object.c [hunk 1]
```diff
@@ -4571,14 +4571,15 @@ mono_unhandled_exception (MonoObject *exc_raw)
...
-	mono_unhandled_exception_checked (exc, &error);
+	mono_unhandled_exception_checked (exc, TRUE, &error);
 	mono_error_assert_ok (&error);
 	HANDLE_FUNCTION_RETURN ();
 }
 
 /**
  * mono_unhandled_exception:
  * @exc: exception thrown
+ * @fatal: if true abort execution
```
Comment (lambdageek) [inline mono/metadata/object.c]: I think we changed to doxygen-style API docs (`\param fatal` etc) at some point (see `mono_unhandled_exception`, above)

File: mcs/class/corlib/System/AppDomain.cs [hunk 2]
```diff
@@ -1459,6 +1462,33 @@ internal byte[] GetMarshalledDomainObjRef ()
...
+			}
+		}
+
+		//XXX This is called by the runtime shutdown process, see runtime.c
+		static bool RunProcessExit ()
+		{
+			object obj = new object ();
+			lock (obj) {
+				ThreadPool.UnsafeQueueUserWorkItem (FireProcessExit, obj);
+				if (!Monitor.Wait (obj, 3000))
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomain.cs]: nit: simple `return Monitor.Wait...` would work too

**Comment by kumpera** [comment]
I looked at the crashes, none are related to shutdown.

**Comment by kumpera** [comment]
unload-appdomain-on-shutdown on linux x64 might be related. I'll investigate.

File: mono/utils/mono-logger-internals.h [hunk 3]
```diff
@@ -28,6 +28,7 @@ typedef enum {
 	MONO_TRACE_IO_LAYER_SEMAPHORE = 1 << 13,
 	MONO_TRACE_IO_LAYER_MUTEX     = 1 << 14,
 	MONO_TRACE_IO_LAYER_HANDLE    = 1 << 15,
+	MONO_TRACE_EXEC               = 1 << 16,
```
Comment (luhenry) [inline mono/utils/mono-logger-internals.h]: What does `exec` stands for?

Comment (kumpera) [reply mono/utils/mono-logger-internals.h]: Misc execution related logging. It's used in this PR to log shutdown messages.

**Comment by kumpera** [comment]
failures look unrelated

**Comment by kumpera** [comment]
@monojenkins build linux aarch64

**Comment by kumpera** [comment]
@monojenkins build linux aarch64 interpreter

**Comment by kumpera** [comment]
@monojenkins build linux x64

**Comment by kumpera** [comment]
@monojenkins build linux x64 FullAOT

**Comment by kumpera** [comment]
@monojenkins build linux x64 mcs

**Comment by kumpera** [comment]
Last commit fix the mcs build.

File: mono/metadata/runtime.c [hunk 4]
```diff
@@ -58,7 +58,8 @@ fire_process_exit_event (MonoDomain *domain, gpointer user_data)
 	ERROR_DECL (error);
 	MonoObject *exc = NULL;
 	MonoMethod *method = mono_class_get_method_from_name (mono_defaults.appdomain_class, "RunProcessExit", -1);
-	g_assert (method);
+	if (!method)
+		return;
```
Comment (akoeplinger) [inline mono/metadata/runtime.c]: why not just bump the corlib version?

Comment (kumpera) [reply mono/metadata/runtime.c]: The runtime needs to be linker/bcl profile resistant.

Comment (akoeplinger) [reply mono/metadata/runtime.c]: I might be missing something but I don't see how this is different to any of the other methods we access from the runtime, that's why we need to keep them via the linker descriptor xml.

Do you mean there are profiles where this method won't exist?

Btw. I just saw that the RunProcessExit() method should be moved before this: 😄
https://github.com/mono/mono/blob/e3cd4e90bfb28e6d599fbac7d6f2989ef5f6dbc0/mcs/class/corlib/System/AppDomain.cs#L1436

File: mcs/class/corlib/System/AppDomain.cs [hunk 5]
```diff
@@ -1459,6 +1462,31 @@ internal byte[] GetMarshalledDomainObjRef ()
...
+					Monitor.Pulse (state);
+				}
+			}
+		}
+
+		//XXX This is called by the runtime shutdown process, see runtime.c
+		static bool RunProcessExit ()
+		{
+			object obj = new object ();
+			lock (obj) {
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomain.cs]: Is this guaranteed to be called only once?

File: mcs/class/corlib/System/AppDomain.cs [hunk 6]
```diff
@@ -1397,6 +1397,9 @@ private void DoDomainUnload ()
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
 		internal extern void DoUnhandledException (Exception e);
 
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		internal extern static void NonFatalUnhandledException (Exception e);
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomain.cs]: Why is it internal ?

File: mcs/class/corlib/System/AppDomain.cs (hunk 5 above)
Comment (kumpera) [reply mcs/class/corlib/System/AppDomain.cs]: Yes, it's called once per domain by the runtime.

File: mcs/class/corlib/System/AppDomain.cs (hunk 6 above)
Comment (kumpera) [reply mcs/class/corlib/System/AppDomain.cs]: To match access of DoUnhandledException

File: mono/metadata/runtime.c (hunk 4 above)
Comment (kumpera) [reply mono/metadata/runtime.c]: It's on the linker files, but there's no downside on making the runtime more resilient on missing optional features.

**Comment by kumpera** [comment]
Moved the AppDomain.cs method to its place. There's nothing else to address from reviewer.

**Comment by kumpera** [comment]
@monojenkins  build

**Comment by kumpera** [comment]
@monojenkins build

**Comment by marek-safar** [comment]
@kumpera there are many `self_interrupt_thread: we can't resume from an async call` when there is finalizer thread could that be related to your change?

**Comment by luhenry** [comment]
There are quite some assertion `self_interrupt_thread: we can't resume from an async call`, @kumpera would you say it's related?

**Comment by luhenry** [comment]
@monojenkins build Windows x64

**Comment by luhenry** [comment]
@monojenkins build Windows i386

**Comment by luhenry** [comment]
@monojenkins build Linux i386

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7

**Comment by kumpera** [comment]
@luhenry I don't think they are related

**Pull request merged**